src/Options.js

```diff
@@ -58,22 +68,69 @@ Options.prototype.extend = function(extension) {
     return new Options(data);
 };
 
+function sizeAtStyle(size, style) {
+    return style.size < 2 ? size : sizeStyleMap[size - 1][style.size - 1];
+}
+
 /**
- * Create a new options object with the given style.
+ * Return an options object with the given style. If `this.style === style`,
+ * returns `this`.
  */
-Options.prototype.withStyle = function(style) {
-    return this.extend({
-        style: style,
-    });
+Options.prototype.havingStyle = function(style) {
+    if (this.style === style) {
+        return this;
+    } else {
+        return this.extend({
+            style: style,
+            size: sizeAtStyle(this.baseSize, style),
+        });
+    }
 };
 
 /**
- * Create a new options object with the given size.
+ * Return an options object with a cramped version of the current style. If
+ * the current style is cramped, returns `this`.
  */
-Options.prototype.withSize = function(size) {
-    return this.extend({
-        size: size,
-    });
+Options.prototype.havingCrampedStyle = function() {
+    return this.havingStyle(this.style.cramp());
+};
+
+/**
+ * Return an options object with the given size and baseSize. If
+ * `this.size === size && this.baseSize === size`, returns `this`.
+ */
+Options.prototype.havingSize = function(size) {
+    if (this.size === size && this.baseSize === size) {
+        return this;
+    } else {
+        // Ensure style is at least `\textstyle`.
+        let style = this.style;
+        if (style.size > 1) {
+            style = style.cramped ? Style.TEXT.cramp() : Style.TEXT;
+        }
+
+        return this.extend({
+            style: style,
+            size: size,
+            baseSize: size,
+        });
+    }
+};
+
+/**
+ * Like `this.havingSize(5).havingStyle(style)`.
```

Comment:
It can create slightly fewer `Options` objects. In particular, imagine `x` has basesize 5 and style SCRIPT. Then `x.havingSize(5).havingStyle(Style.SCRIPT) != x`, but `x.havingBaseStyle(Style.SCRIPT) == x`.

The function was more useful (i.e., did more than just optimize) in a prior version, before I decided that it was best for `.havingSize()` to reset the style.